connaisseur/alert.py

```diff
@@ -154,58 +177,37 @@ def __init__(self, alert_receiver_config: dict):
                         f"An error occurred while loading the token file {token_file}: {str(err)}"
                     )
 
-            self.authorization_prefix = bearer_authentication_config.get(
-                "authorization_prefix", "Bearer"
-            )
-            # TODO maybe validate authorization prefix
-
         def get_header(self) -> dict:
-            return {"Authorization": f"{self.authorization_prefix} {self.token}"}
+            return {"Authorization": f"{self.authentication_scheme} {self.token}"}
+
+    init_map = {
+        "basic": AlertReceiverBasicAuthentication,
+        "bearer": AlertReceiverBearerAuthentication,
+        "none": AlertReceiverNoneAuthentication,
+    }
+
+    _authentication_instance = None
 
     def __init__(self, alert_receiver_config: dict):
         self.authentication_type = alert_receiver_config.get(
             "receiver_authentication_type", "none"
         )
+        self.__init_authentication_instance(alert_receiver_config)
 
-        if self.is_basic():
-            self.__init_basic_authentication(alert_receiver_config)
-        elif self.is_bearer():
-            self.__init_bearer_authentication(alert_receiver_config)
+    def __init_authentication_instance(self, alert_receiver_config: dict):
+        authentication_class = self.__get_authentication_class()
+        self._authentication_instance = authentication_class(alert_receiver_config)
 
-    def is_basic(self):
-        return self.authentication_type == "basic"
-
-    def is_bearer(self):
-        return self.authentication_type == "bearer"
+    def __get_authentication_class(self):
+        if self.authentication_type not in AlertReceiverAuthentication.init_map.keys():
+            raise ConfigurationError(
+                f"No authentication type found. Valid values are {list(AlertReceiverAuthentication.init_map.keys())}"
+            )  # hopefully this never happens
 
-    def is_none(self):
-        return self.authentication_type == "none"
-
-    def __init_bearer_authentication(self, alert_receiver_config: dict):
-        self.bearer_authentication = (
-            AlertReceiverAuthentication.AlertReceiverBearerAuthentication(
-                alert_receiver_config
-            )
-        )
-
-    def __init_basic_authentication(self, alert_receiver_config: dict):
-        self.basic_authentication = (
-            AlertReceiverAuthentication.AlertReceiverBasicAuthentication(
-                alert_receiver_config
-            )
-        )
+        return self.init_map.get(self.authentication_type)
```

Comment:
```python
try:
    return self.init_map[self.authentication_type]
except KeyError:
    raise ConfigurationError(...)
```

This is a Python paradigm called EAFP. For context, see https://sauravmaheshkar.github.io/blog/Introduction-To-Duck-Typing-and-EAFP/.

One benefit is that EAFP gets ride of many [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) race conditions (although not the case here, as the state is purely local).

Reply:
I was not aware of this python principle. Thank you.